From e2bd08084a47b9830a1443e712e27e12b24664a6 Mon Sep 17 00:00:00 2001
From: Hal Gentz <zegentzy@protonmail.com>
Date: Sun, 15 Sep 2019 15:29:50 -0600
Subject: [PATCH 2/3] egl: Puts RGBA visuals in the second config selection
 group.

That way applications don't get windows that are compositor alpha-blended
accidentally.

In the ideal world, this would be done by the xserver, as it does for
GLX, however, an appropriate place could not be found, so it's being
placed here instead.

Fixes: 049f343e8ac "egl: Allow 24-bit visuals for 32-bit RGBA8888 configs"
Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=67676
Cc: mesa-stable@lists.freedesktop.org
Signed-off-by: Hal Gentz <zegentzy@protonmail.com>
---
 src/egl/drivers/dri2/platform_x11.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/egl/drivers/dri2/platform_x11.c b/src/egl/drivers/dri2/platform_x11.c
index 9cef0e0b35a..8e43c659a8a 100644
--- a/src/egl/drivers/dri2/platform_x11.c
+++ b/src/egl/drivers/dri2/platform_x11.c
@@ -883,9 +883,17 @@ dri2_x11_add_configs_for_visuals(struct dri2_egl_display *dri2_dpy,
             dri2_conf = dri2_add_config(disp, config, config_count + 1,
                                         surface_type, config_attrs,
                                         rgba_shifts, rgba_sizes);
-            if (dri2_conf)
+            if (dri2_conf) {
                if (dri2_conf->base.ConfigID == config_count + 1)
                   config_count++;
+
+               /* Put RGBA visuals in the second ConfigSelectGroup so that they
+                * have lower priority. Applications probably don't want the
+                * compositor to alpha-blend their windows.
+                */
+               if (d.data->depth != 24 && d.data->depth != 30)
+                  dri2_conf->base.ConfigSelectGroup = 1;
+            }
          }
       }
 
-- 
2.23.0

